elytron/WFCORE-3596-named-permission-sets.adoc

```diff
@@ -0,0 +1,110 @@
+  removed. Without named permission sets, the provisioning tool would not know which permission mapping a subsystem
+  permission should be added to or removed from. Note that the provisioning tool will only be touching the "default-permissions".
+  It won't attempt to touch any user-specified permission sets.
+
+[source,xml]
+----
+<mappers>
+...
+    <simple-permission-mapper name="default-permission-mapper" mapping-mode="first">
+        <permission-mapping>
+            <principal name="anonymous"/>
+            <permission-set name="default-permissions"/>
+        </permission-mapping>
+        <permission-mapping match-all="true">
+            <permission-set name="login-permission" />
+            <permission-set name="default-permissions"/>
+        </permission-mapping>
+    </simple-permission-mapper>
+...
+</mappers>
+
+<permission-sets>
+    <permission-set name="login-permission">
+        <permission class-name="org.wildfly.security.auth.permission.LoginPermission"/>
+    </permission-set>
+    <permission-set name="default-permissions">
+        <permission class-name="org.wildfly.extension.batch.jberet.deployment.BatchPermission" module="org.wildfly.extension.batch.jberet" target-name="*"/>
+        <permission class-name="org.wildfly.transaction.client.RemoteTransactionPermission" module="org.wildfly.transaction.client"/>
+        <permission class-name="org.jboss.ejb.client.RemoteEJBPermission" module="org.jboss.ejb-client"/>
+    </permission-set>
+</permission-sets>
+----
+
+* A permission set should be a private capability.
+
+* When transforming the model for legacy slaves, the permission sets should be in-lined, i.e., a permission set that's
+  referenced from a permission mapping or from a constant permission mapper would just be replaced with the permissions
+  that make up the permission set. For example, the "default-permissions" permission set referenced from the permission
+  mappings in the above example would be transformed to:
+
+[source,xml]
+----
+<permission class-name="org.wildfly.extension.batch.jberet.deployment.BatchPermission" module="org.wildfly.extension.batch.jberet" target-name="*"/>
+<permission class-name="org.wildfly.transaction.client.RemoteTransactionPermission" module="org.wildfly.transaction.client"/>
+<permission class-name="org.jboss.ejb.client.RemoteEJBPermission" module="org.jboss.ejb-client"/>
+----
+
+* XML configuration for prior versions should be automatically converted to the new version, i.e., each list of permissions should be
+  converted to a permission set. In particular, each list of permissions in a permission mapping or constant permission mapper
+  should be automatically replaced with a reference to a newly created permission set that is made up of the list of permissions.
+  The name of the permission set could be a combination of the mapper name and the class-name of each permission in the list.
```

Comment:
But that won't help to provisioning tool which will manipulate only set named "default-permissions"

Reply:
@darranl Do you have any thoughts on this strategy for converting XML configuration for prior versions to the new version with named permission sets?

Reply:
The default permission set consists of the subsystem permissions only. Collecting all the permissions from all the mappings wouldn't be right as this would also include ```org.wildfly.security.auth.permission.LoginPermission``` (and also any other user-specified permissions from other mappings), which aren't meant to be part of the default permission set.

Reply:
That's good to know, thanks. Slightly OT, @fjuma you mentioned a class name as an element of a permission set name (which i assume has to be unique). Would it be safe to assume that the class name would work as a permission id?

Reply:
My initial thinking was that class name would work as a permission id. But, technically, one could have permissions with the same class name but different modules. So the combination of the class name and the module is likely better as a permission id.